Source/WebCore/css/StyleProperties.cpp

```diff
@@ -413,179 +547,163 @@ static std::optional<CSSValueID> fontStretchKeyword(double value)
+    auto path = includePath ? serializeLonghandValue(pathIndex) : String();
+    auto distanceSeparator = includeDistance ? " " : "";
+    auto distance = includeDistance ? serializeLonghandValue(distanceIndex) : String();
+    auto rotateSeparator = includeRotate ? " " : "";
+    auto rotate = includeRotate ? serializeLonghandValue(rotateIndex) : String();
+    auto anchorSeparator = includeAnchor ? " / " : "";
+    auto anchor = includeAnchor ? serializeLonghandValue(anchorIndex) : String();
+
+    return makeString(position,
+        pathSeparator, path,
+        distanceSeparator, distance,
+        rotateSeparator, rotate,
+        anchorSeparator, anchor);
+}
+
+String ShorthandSerializer::serializeFontVariant() const
+{
+    for (auto& value : longhandValues()) {
+        if (CSSPropertyParserHelpers::isSystemFontShorthand(valueID(value)))
+            return String();
+    }
+    if (isLonghandValueNone(longhandIndex(0, CSSPropertyFontVariantLigatures))) {
+        for (auto longhand : longhands()) {
+            // font-variant cannot represent "font-variant-ligatures: none" along with any other non-normal longhands.
+            if (longhand.property != CSSPropertyFontVariantLigatures && !isInitialValue(longhand))
+                return String();
+        }
     }
-    return result.isEmpty() ? nameString(CSSValueNormal) : result.toString();
+    return serializeLonghandsOmittingInitialValues();
 }
 
-String StyleProperties::fontSynthesisValue() const
+String ShorthandSerializer::serializeFontSynthesis() const
 {
     // font-synthesis: none | [ weight || style || small-caps ]
-    ASSERT(fontSynthesisShorthand().length() == 3);
-    auto weight = propertyAsValueID(CSSPropertyFontSynthesisWeight).value_or(CSSValueInvalid);
-    auto style = propertyAsValueID(CSSPropertyFontSynthesisStyle).value_or(CSSValueInvalid);
-    auto caps = propertyAsValueID(CSSPropertyFontSynthesisSmallCaps).value_or(CSSValueInvalid);
+    ASSERT(length() == 3);
 
-    if (weight == CSSValueNone && style == CSSValueNone && caps == CSSValueNone)
+    bool weight = !isLonghandValueNone(longhandIndex(0, CSSPropertyFontSynthesisWeight));
+    bool style = !isLonghandValueNone(longhandIndex(1, CSSPropertyFontSynthesisStyle));
+    bool caps = !isLonghandValueNone(longhandIndex(2, CSSPropertyFontSynthesisSmallCaps));
+
+    if (!weight && !style && !caps)
         return nameString(CSSValueNone);
 
-    StringBuilder result;
-    if (weight == CSSValueAuto)
-        result.append("weight"_s);
-    if (style == CSSValueAuto)
-        result.append(result.isEmpty() ? "" : " ", "style"_s);
-    if (caps == CSSValueAuto)
-        result.append(result.isEmpty() ? "" : " ", "small-caps"_s);
-    return result.toString();
+    return makeString(
+        !weight ? "" : "weight",
+        !weight || !style ? "" : " ",
+        !style ? "" : "style",
+        !style || !caps ? "" : " ",
+        !caps ? "" : "small-caps"
```

Comment:
The separator is missing for `weight = true, style = false, caps = true`.

```suggestion
        weight ? "weight" : "",
        weight && (style || caps) ? " " : "",
        style ? "style" : "",
        style && caps ? " " : "",
        caps ? "small-caps" : ""
```

Reply:
Yes, definitely a mistake and I'll do something like that. I am also considering just using a table of the 7 serializations since there are only three Booleans here.